python/ray/data/datasource/file_datasink.py

```diff
@@ -189,13 +186,15 @@ def write_block(self, block: BlockAccessor, block_index: int, ctx: TaskContext):
             )
             write_path = posixpath.join(self.path, filename)
 
-            def write_row_to_path():
+            def write_row_to_path(row, write_path):
                 with self.open_output_stream(write_path) as file:
                     self.write_row_to_file(row, file)
 
             logger.debug(f"Writing {write_path} file.")
             call_with_retry(
-                write_row_to_path,
+                lambda row=row, write_path=write_path: write_row_to_path(
+                    row, write_path
+                ),
```

Comment:
unrelated to this PR, but fixing the following lint error
`python/ray/data/datasource/file_datasink.py:190:46: B023 Function definition does not bind loop variable 'write_path'.`